File: test/ui/g2d.js
/**
 * R2D 2D 绘制综合示例
 *
 * 覆盖所有 Phase 1 API：
 *   fillRect / strokeRect / clearRect
 *   beginPath / moveTo / lineTo / quadraticCurveTo / bezierCurveTo
 *   arc / ellipse / closePath
 *   fill / stroke
 *   save / restore
 *   setFillStyle / setStrokeStyle / setLineWidth / setGlobalAlpha
 */
import { View, G2D, Root } from 'kwikui';

// ─────────────────────────────────────────────────
// 创建 R2D 实例（画布 400×300 逻辑像素）
// ─────────────────────────────────────────────────
const g2d = G2D({ id: "demo", width: 400, height: 300 });

// ═════════════════════════════════════════════════
//  1. fillRect — 填充矩形
//     setFillStyle — 设置填充颜色（支持 #rgb / #rrggbb / rgba）
// ═════════════════════════════════════════════════

// 红色实心矩形（左下角）
g2d.setFillStyle("#E74C3C");
g2d.fillRect(10, 10, 120, 80);

// 橙色实心矩形（紧邻右侧）
g2d.setFillStyle("#F39C12");
g2d.fillRect(140, 10, 120, 80);

// 黄色实心矩形（使用 rgba 格式）
g2d.setFillStyle("rgba(241,196,15,1)");
g2d.fillRect(250, 10, 120, 80);   // 逐步右移


// ═════════════════════════════════════════════════
//  2. strokeRect — 描边矩形（仅边框）
//     setStrokeStyle → 描边颜色
//     setLineWidth   → 线宽
// ═════════════════════════════════════════════════

// 绿色粗边框矩形
g2d.setStrokeStyle("#2ECC71");
g2d.setLineWidth(4);
g2d.strokeRect(10, 110, 120, 80);

// 蓝色中等边框矩形
g2d.setStrokeStyle("#3498DB");
g2d.setLineWidth(2);
g2d.strokeRect(140, 110, 120, 80);

// 紫色细边框矩形
g2d.setStrokeStyle("#9B59B6");
g2d.setLineWidth(1);
g2d.strokeRect(270, 110, 120, 80);

// ═════════════════════════════════════════════════
//  3. 直线路径 — beginPath / moveTo / lineTo / closePath
//     fill → 填充路径区域
// ═════════════════════════════════════════════════

// 红色填充三角形
g2d.setFillStyle("#E74C3C");
g2d.beginPath();
g2d.moveTo(20, 220);       // 起点
g2d.lineTo(90, 280);       // 连线到右下
g2d.lineTo(20, 280);       // 连线到左下
g2d.closePath();           // 闭合路径（回到起点）
g2d.fill();                // 填充路径内部

// ═════════════════════════════════════════════════
//  4. 二次贝塞尔曲线 — quadraticCurveTo / stroke
// ═════════════════════════════════════════════════

// 绿色开口向上的抛物线
g2d.setStrokeStyle("#2ECC71");
g2d.setLineWidth(3);
g2d.beginPath();
g2d.moveTo(130, 280);             // 起点（左下）
g2d.quadraticCurveTo(180, 220,    // 控制点（顶点处）
                      230, 280);  // 结束点（右下）
g2d.stroke();

// ═════════════════════════════════════════════════
//  5. 三次贝塞尔曲线 — bezierCurveTo / stroke
// ═════════════════════════════════════════════════

// 蓝色 S 形曲线
g2d.setStrokeStyle("#3498DB");
g2d.setLineWidth(3);
g2d.beginPath();
g2d.moveTo(260, 280);
g2d.bezierCurveTo(290, 210,   // 控制点 1（上方，拉出右弯）
                   330, 310,   // 控制点 2（下方，拉回左弯）
                   350, 260);  // 结束点（右上方，形成 S 出口）
g2d.stroke();

// ═════════════════════════════════════════════════
//  6. 圆弧 — arc 完整圆 (fill) + 半圆弧 (stroke)
//     arc(cx, cy, r, startAngle, endAngle, ccw)
// ═════════════════════════════════════════════════

// 青色填充圆形
g2d.setFillStyle("#1ABC9C");
g2d.beginPath();
g2d.arc(280, 210, 30,        // 圆心 (280,210), 半径 30
        0, Math.PI * 2,       // 0 → 2π（完整圆）
        false);
g2d.fill();

// 橙色描边半圆弧（拱形）
g2d.setStrokeStyle("#E67E22");
g2d.setLineWidth(4);
g2d.beginPath();
g2d.arc(340, 240, 28,        // 圆心 (340,240), 半径 28
        0, Math.PI,           // 0 → π（上半弧）
        false);
g2d.stroke();

// ═════════════════════════════════════════════════
//  7. 椭圆 — ellipse / fill
//     ellipse(cx, cy, rx, ry, rotation, startAngle, endAngle, ccw)
// ═════════════════════════════════════════════════

// 紫色填充椭圆（旋转 30°）
g2d.setFillStyle("#8E44AD");
g2d.beginPath();
g2d.ellipse(80, 210,         // 中心 (80,210)
            45, 20,           // 半长轴 45, 半短轴 20
            0.5,              // 旋转 0.5 弧度 ≈ 30°
            0, Math.PI * 2,   // 完整椭圆
            false);
g2d.fill();

// ═════════════════════════════════════════════════
//  8. save / restore + globalAlpha
//     save → 保存当前样式状态入栈
//     restore → 从栈顶恢复样式状态
//     setGlobalAlpha → 全局透明度 (0.0 ~ 1.0)
// ═════════════════════════════════════════════════

// 保存当前状态（fillStyle=紫色, strokeStyle=橙色, lineWidth=4, alpha=1.0）
g2d.save();

// 降低透明度绘制两个半透明矩形（相互叠加）
g2d.setGlobalAlpha(0.4);
g2d.setFillStyle("#E74C3C");
g2d.fillRect(150, 200, 60, 60);
g2d.setFillStyle("#3498DB");
g2d.fillRect(180, 220, 60, 60);

// 恢复之前的状态（fillStyle 恢复为紫色，alpha 恢复为 1.0）
g2d.restore();

// 用恢复后的紫色画一个小圆点确认 restore 生效
g2d.beginPath();
g2d.arc(380, 60, 15, 0, Math.PI * 2, false);
g2d.fill();

// ═════════════════════════════════════════════════
//  9. clearRect — 清除矩形区域（挖空效果）
//     清除区域内的所有绘制内容变为透明
// ═════════════════════════════════════════════════

// 从第一个红色矩形中挖掉一块
g2d.clearRect(30, 20, 60, 50);

// 从黄色矩形中挖掉一块
g2d.clearRect(290, 30, 80, 40);

// =================================================
// 导出：放入 Root 容器，20px 内边距
// =================================================
export default Root(
    View({ padding: 20, width: 400, height: 400 }, [g2d])
);

File: src/bridge/g2d.cpp
/** @brief G2D 元素: JS 工厂 / 方法 / 插件式注册 (自 bindings.cpp / element_parser.cpp 抽取)。 */
module;
#include "quickjs.h"
#include <atomic>

module kwik.bridge.g2d;

import kwik.element.g2d;           // G2D 类
import kwik.element.view;          // View (findById / unique_ptr<View>)
import kwik.element.typed_prop;    // TypedPropMap
import kwik.bridge.props_parser;   // PropsExtractor / parseViewProps
import kwik.bridge.bindings;       // makeElementHelper
import kwik.bridge.element_spec;   // ElementSpec
import kwik.bridge.element_parser; // ElementParser::registerExtension / applyBindings
import kwik.engine.context;        // QuickJSContext (getUserPointer)
import kwik.core.color_parser;     // parseColor (setFillStyle / setStrokeStyle)
import kwik.engine.js_value;

import std;

// ── 辅助宏：提取 G2D* 指针 (自 bindings.cpp:1041 原样搬迁) ──
#define G2D_FROM_THIS(ctx, this_val)                                                                                   \
    [&]() -> G2D * {                                                                                                   \
        JSValue _p = JS_GetPropertyStr(ctx, this_val, "props");                                                        \
        /* 优先读 __g2d_ptr（eager 创建 or 首次 findById 后缓存） */                                                   \
        JSValue _ptr = JS_GetPropertyStr(ctx, _p, "__g2d_ptr");                                                        \
        if (!JS_IsUndefined(_ptr)) {                                                                                   \
            double _v;                                                                                                 \
            JS_ToFloat64(ctx, &_v, _ptr);                                                                              \
            JS_FreeValue(ctx, _ptr);                                                                                   \
            JS_FreeValue(ctx, _p);                                                                                     \
            return reinterpret_cast<G2D *>(static_cast<uintptr_t>(_v));                                                \
        }                                                                                                              \
        JS_FreeValue(ctx, _ptr);                                                                                       \
        /* Fallback: findById（兼容树已建好但没有缓存的场景） */                                                       \
        JSValue _id = JS_GetPropertyStr(ctx, _p, "id");                                                                \
        const char *_idStr = JS_ToCString(ctx, _id);                                                                   \
        auto *_qctx = static_cast<QuickJSContext *>(JS_GetContextOpaque(ctx));                                         \
        View *_root = static_cast<View *>(_qctx->getUserPointer());                                                    \
        auto *_r = dynamic_cast<G2D *>(_root ? _root->findById(_idStr) : nullptr);                                     \
        if (_r) { /* 缓存到 __g2d_ptr 下次直接走快路径 */                                                              \
            JS_SetPropertyStr(ctx, _p, "__g2d_ptr",                                                                    \
                              JS_NewFloat64(ctx, static_cast<double>(reinterpret_cast<uintptr_t>(_r))));               \
        }                                                                                                              \
        JS_FreeCString(ctx, _idStr);                                                                                   \
        JS_FreeValue(ctx, _id);                                                                                        \
        JS_FreeValue(ctx, _p);                                                                                         \
        return _r;                                                                                                     \
    }()

// ── 各方法回调 ──
static JSValue js_g2d_fillRect(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    g2d->fillRect(f(0), f(1), f(2), f(3));
    return JS_UNDEFINED;
}

static JSValue js_g2d_strokeRect(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    g2d->strokeRect(f(0), f(1), f(2), f(3));
    return JS_UNDEFINED;
}

static JSValue js_g2d_clearRect(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    g2d->clearRect(f(0), f(1), f(2), f(3));
    return JS_UNDEFINED;
}

static JSValue js_g2d_beginPath(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    g2d->beginPath();
    return JS_UNDEFINED;
}

static JSValue js_g2d_moveTo(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    g2d->moveTo(f(0), f(1));
    return JS_UNDEFINED;
}

static JSValue js_g2d_lineTo(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    g2d->lineTo(f(0), f(1));
    return JS_UNDEFINED;
}

static JSValue js_g2d_quadraticCurveTo(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    g2d->quadraticCurveTo(f(0), f(1), f(2), f(3));
    return JS_UNDEFINED;
}

static JSValue js_g2d_bezierCurveTo(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    g2d->bezierCurveTo(f(0), f(1), f(2), f(3), f(4), f(5));
    return JS_UNDEFINED;
}

static JSValue js_g2d_arc(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    bool ccw = argc > 5 ? JS_ToBool(ctx, argv[5]) != 0 : false;
    g2d->arc(f(0), f(1), f(2), f(3), f(4), ccw);
    return JS_UNDEFINED;
}

static JSValue js_g2d_ellipse(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    bool ccw = argc > 7 ? JS_ToBool(ctx, argv[7]) != 0 : false;
    g2d->ellipse(f(0), f(1), f(2), f(3), f(4), f(5), f(6), ccw);
    return JS_UNDEFINED;
}

static JSValue js_g2d_closePath(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    g2d->closePath();
    return JS_UNDEFINED;
}

static JSValue js_g2d_fill(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    g2d->fill();
    return JS_UNDEFINED;
}

static JSValue js_g2d_stroke(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    g2d->stroke();
    return JS_UNDEFINED;
}

static JSValue js_g2d_clip(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    g2d->clip();
    return JS_UNDEFINED;
}

static JSValue js_g2d_save(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    g2d->save();
    return JS_UNDEFINED;
}

static JSValue js_g2d_restore(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    g2d->restore();
    return JS_UNDEFINED;
}

static JSValue js_g2d_drawImage(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    double texId;
    JS_ToFloat64(ctx, &texId, argv[0]);
    auto f = [&](int i) {
        double v;
        JS_ToFloat64(ctx, &v, argv[i]);
        return float(v);
    };
    g2d->drawImage(uint32_t(texId), f(1), f(2), f(3), f(4));
    return JS_UNDEFINED;
}

static JSValue js_g2d_setFillStyle(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    const char *s = JS_ToCString(ctx, argv[0]);
    if (s) {
        g2d->setFillStyle(parseColor(s));
        JS_FreeCString(ctx, s);
    }
    return JS_UNDEFINED;
}

static JSValue js_g2d_setStrokeStyle(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    const char *s = JS_ToCString(ctx, argv[0]);
    if (s) {
        g2d->setStrokeStyle(parseColor(s));
        JS_FreeCString(ctx, s);
    }
    return JS_UNDEFINED;
}

static JSValue js_g2d_setLineWidth(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    double v;
    JS_ToFloat64(ctx, &v, argv[0]);
    g2d->setLineWidth(float(v));
    return JS_UNDEFINED;
}

static JSValue js_g2d_setGlobalAlpha(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    double v;
    JS_ToFloat64(ctx, &v, argv[0]);
    g2d->setGlobalAlpha(float(v));
    return JS_UNDEFINED;
}

static JSValue js_g2d_reset(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
    auto *g2d = G2D_FROM_THIS(ctx, this_val);
    if (!g2d) return JS_UNDEFINED;
    g2d->reset();
    return JS_UNDEFINED;
}

// ── bindG2DMethods (自 bindings.cpp:1291 原样搬迁) ──
static void bindG2DMethods(JSContext *ctx, JSValue obj) {
    struct {
        const char *name;
        JSCFunction *fn;
        int argc;
    } methods[] = {
        {"fillRect", js_g2d_fillRect, 4},
        {"strokeRect", js_g2d_strokeRect, 4},
        {"clearRect", js_g2d_clearRect, 4},
        {"beginPath", js_g2d_beginPath, 0},
        {"moveTo", js_g2d_moveTo, 2},
        {"lineTo", js_g2d_lineTo, 2},
        {"quadraticCurveTo", js_g2d_quadraticCurveTo, 4},
        {"bezierCurveTo", js_g2d_bezierCurveTo, 6},
        {"arc", js_g2d_arc, 6},
        {"ellipse", js_g2d_ellipse, 8},
        {"closePath", js_g2d_closePath, 0},
        {"fill", js_g2d_fill, 0},
        {"stroke", js_g2d_stroke, 0},
        {"clip", js_g2d_clip, 0},
        {"save", js_g2d_save, 0},
        {"restore", js_g2d_restore, 0},
        {"drawImage", js_g2d_drawImage, 5},
        {"reset", js_g2d_reset, 0},
        {"setFillStyle", js_g2d_setFillStyle, 1},
        {"setStrokeStyle", js_g2d_setStrokeStyle, 1},
        {"setLineWidth", js_g2d_setLineWidth, 1},
        {"setGlobalAlpha", js_g2d_setGlobalAlpha, 1},
    };
    for (auto &m : methods) { JS_SetPropertyStr(ctx, obj, m.name, JS_NewCFunction(ctx, m.fn, m.name, m.argc)); }
}

// ── JS 工厂 (自 bindings.cpp:1323 搬迁, 仅 makeElement → makeElementHelper) ──
static JSValue js_g2d(JSContext *ctx, JSValueConst, int argc, JSValueConst *argv) {
    JSValue props = (argc > 0 && JS_IsObject(argv[0])) ? argv[0] : JS_UNDEFINED;
    if (JS_IsObject(props)) {
        JSValue idVal = JS_GetPropertyStr(ctx, props, "id");
        if (JS_IsUndefined(idVal)) {
            static std::atomic<uint32_t> g2d_id_counter{0};
            char buf[64];
            std::snprintf(buf, sizeof(buf), "__g2d_%u", g2d_id_counter++);
            JS_SetPropertyStr(ctx, props, "id", JS_NewString(ctx, buf));
        }
        JS_FreeValue(ctx, idVal);
    }
    JSValue obj = makeElementHelper(ctx, "G2D", props, JS_UNDEFINED);    // 改这里
    bindG2DMethods(ctx, obj);

    // 立即创建 C++ G2D, 指针存到 props (命令式绘制在树建立前就要到达 C++)
    auto *g2d = new G2D();
    JS_SetPropertyStr(ctx, props, "__g2d_ptr",
                      JS_NewFloat64(ctx, static_cast<double>(reinterpret_cast<uintptr_t>(g2d))));
    return obj;
}

// ── 创建器 (自 element_parser.cpp:444-465 搬迁) ──
static std::unique_ptr<View> createG2D(const JSValueRef &pv) {
    // 复用 eager 创建的 C++ G2D
    if (pv.hasProperty("__g2d_ptr")) {
        auto ptrVal = pv.getProperty("__g2d_ptr");
        JSContext *ctx = pv.context();
        double v;
        JS_ToFloat64(ctx, &v, ptrVal.raw());
        auto *existing = reinterpret_cast<G2D *>(static_cast<uintptr_t>(v));
        TypedPropMap meta;
        PropsExtractor ex(pv, &meta);
        existing->props = parseViewProps(ex);     // 更新 ViewProps (width/height 等)
        return std::unique_ptr<G2D>(existing);    // 树接管所有权
    }
    // 降级: 无 eager 场景正常创建
    TypedPropMap meta;
    PropsExtractor ex(pv, &meta);
    auto v = std::make_unique<G2D>(parseViewProps(ex));
    v->propMeta = std::move(meta);
    applyBindings(v.get(), pv);
    return v;
}

// ── 自注册 ──
void registerG2DElement() {
    ElementSpec spec;
    spec.typeName = "G2D";
    spec.creator = createG2D;
    // reconcileProps / attachHandlers 无需: G2D 无专有 reconcile / 事件形状
    spec.jsFactoryName = "G2D";
    spec.jsFactoryFn = js_g2d;
    spec.jsFactoryArgc = 1;
    ElementParser::registerExtension(std::move(spec));
}